Rdmp.Dicom/Cache/Pipeline/SMICacheSource.cs

```diff
@@ -132,7 +132,13 @@ protected DicomCFindRequest CreateStudyRequestByDateRangeForModality(DateTime da
             request.Dataset.AddOrUpdate(DicomTag.ModalitiesInStudy, modalityToQuery);
             var studyDateRange = new DicomDateRange(dateFrom, dateTo);
             request.Dataset.AddOrUpdate(DicomTag.StudyDate, studyDateRange);
-            request.Dataset.AddOrUpdate(DicomTag.StudyTime, studyDateRange);
+
+            // Problem: time and date are orthogonal in DICOM, so '9:00 1/1/20 to 10:00 31/1/20' can
+            // skip data from e.g. 15:00 2/2/20 because it isn't '9:00-10:00'.
+            // To avoid this, ignore time values unless we're doing less than a whole day:
+            if ((dateTo - dateFrom).Days == 0)
```

Comment:
Does date subtraction return a timespan? Is 11pm to 1am the next day would be Days == 1 or Days == 0?

Reply:
Can you write some tests for this (e.g. by subclassing SMICacheSource and calling the protected method (or make it public tbh it doesn't need to be protected).

Reply:
> Does date subtraction return a timespan? Is 11pm to 1am the next day would be Days == 1 or Days == 0?

01/02/2020 01:00:00-01/01/2020 23:00:00 = 02:00:00 (days 0, hours 2)

Reply:
> In practice this shouldn't have affected any of our data because up to now we are only pulling 1 hour periods and the minutes/seconds is always 00.

Doesn't affect the live Tayside PACS feed, because that only fetches 1 hour windows anyway, but was silently defeating the unit tests in various exciting ways.

Reply:
> Can you write some tests for this (e.g. by subclassing SMICacheSource and calling the protected method (or make it public tbh it doesn't need to be protected).
> 
> i.e. some
> 
> ```cs
> [TestCase('2001-01-01 11:00:00','2001-01-02 11:00:00',false]
> public void TestCFindDateRange(string start, string end, expectTimeToBeIncludedInCFind)
> ```

Erk. Actually that exposes a flaw in the current workaround: because the dates get split from times, that pair wouldn't fetch the 2 hour window (11pm 1/1 to 1am 2/1), but the 46 hour window (1am 1/1 to 11pm 2/1). Should the method throw for date/time ranges that don't map directly to unambiguous DICOM ranges? The alternative is probably to break it into multiple fetches (so 11pm-midnight on 1/1 then midnight-1am on 2/1).

Reply:
Something like this I think plays best with the caching engine

```csharp
DateTime dt1;
DateTime dt2;


if(dt1.Date != dt2.Date)
{
    if(dt1.TimeOfDay != TimeSpan.Zero || dt2.TimeOfDay < new TimeSpan(23,59,59))
    {
        throw new ArgumentOutOfRangeException("When creating multi day fetch windows.  The time range must be from 00:00:00 on the start date and 23:59:59 on the ending day")
    }
}
```

Reply:
Actually as it stands this PR will break the existing system for the fetch window 11pm to 00:00:00 (of next day)

Reply:
No, for that 1 hour window the condition is false so it'll set exactly the same parameters it does now. (Those parameters might not be _right_ at the moment - if we ask for "11:00pm day 1 to 0:00 day 2" we're actually asking for two different one-hour windows, the last hour of _both_ days, not just the first one - but the current PR doesn't change that: it only suppresses the time filtering for windows of 24 hours or more.)